Conversions - Length: Adding new units

scottwarren opened the pull request:
## Description of new Instant Answer, or changes
Adds suggested length units to the length Instant Answers API


## Related Issues and Discussions
Fixes #4175 

Defined in Math.js:

* [x] Micron
* [x] Femtometer
* [x] Attometer
* [x] zeptometer
* [x] Yoctometer
* [x] Dekameter
* [x] Hectometer
* [x] Metameter
* [x] Gigameter
* [x] Terameter
* [x] Petameter
* [x] Exameter
* [x] Zettameter
* [x] Yottameter
* [x] Nautical Mile
* [x] Li
* [x] Half Marathon
* [x] Marathon
* [x] Parsec
* [x] Milliparsec
* [x] Nanoparsec
* [x] Picoparsec
* [x] Kiloparsec
* [x] Metaparsec
* [x] Gigaparsec
* [x] Teraparsec
* [x] Astronomical Unit
* [x] Light Year
* [x] League
* [x] Fathom
* [x] Smoot
* [x] Cubit
* [x] Furlong
* [x] Megafurlong
* [x] Beard Second
* [ ] Zeta

Defined in Triggers (with aliases)

* [x] Micron
* [x] Femtometer
* [x] Attometer
* [x] zeptometer
* [x] Yoctometer
* [x] Dekameter
* [x] Hectometer
* [x] Metameter
* [x] Gigameter
* [x] Terameter
* [x] Petameter
* [x] Exameter
* [x] Zettameter
* [x] Yottameter
* [x] Nautical Mile
* [x] Li
* [x] Half Marathon
* [x] Marathon
* [x] Parsec
* [x] Milliparsec
* [x] Nanoparsec
* [x] Picoparsec
* [x] Kiloparsec
* [x] Metaparsec
* [x] Gigaparsec
* [x] Teraparsec
* [x] Astronomical Unit
* [x] Light Year
* [x] League
* [x] Fathom
* [x] Smoot
* [x] Cubit
* [x] Furlong
* [x] Megafurlong
* [x] Beard Second
* [ ] Zeta

Tested (frontend)

* [x] Micron
* [x] Femtometer
* [x] Attometer
* [x] zeptometer
* [x] Yoctometer
* [x] Dekameter
* [x] Hectometer
* [x] Metameter
* [x] Gigameter
* [x] Terameter
* [x] Petameter
* [x] Exameter
* [x] Zettameter
* [x] Yottameter
* [x] Nautical Mile
* [x] Li
* [x] Half Marathon
* [x] Marathon
* [x] Parsec
* [x] Milliparsec
* [x] Nanoparsec
* [x] Picoparsec
* [x] Kiloparsec
* [x] Metaparsec
* [x] Gigaparsec
* [x] Teraparsec
* [x] Astronomical Unit
* [x] Light Year
* [x] League
* [x] Fathom
* [x] Smoot
* [x] Cubit
* [x] Furlong
* [x] Megafurlong
* [x] Beard Second
* [ ] Zeta

Tested (backend):

* [x] Micron
* [x] Femtometer
* [x] Attometer
* [x] zeptometer
* [x] Yoctometer
* [x] Dekameter
* [x] Hectometer
* [x] Metameter
* [x] Gigameter
* [x] Terameter
* [x] Petameter
* [x] Exameter
* [x] Zettameter
* [x] Yottameter
* [x] Nautical Mile
* [x] Li
* [x] Half Marathon
* [x] Marathon
* [x] Parsec
* [x] Milliparsec
* [x] Nanoparsec
* [x] Picoparsec
* [x] Kiloparsec
* [x] Metaparsec
* [x] Gigaparsec
* [x] Teraparsec
* [x] Astronomical Unit
* [x] Light Year
* [x] League
* [x] Fathom
* [x] Smoot
* [x] Cubit
* [x] Furlong
* [x] Megafurlong
* [x] Beard Second
* [ ] Zeta

---

<!-- The Instant Answer ID can be found by clicking the `?` icon beside the Instant Answer result on DuckDuckGo.com -->
Instant Answer Page: https://duck.co/ia/view/conversions
<!-- FILL THIS IN:                           ^^^^ -->

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

scottwarren commented:
# Pull request testing:

## People to notify
<!-- Please @mention any relevant people/organizations here: -->


## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting
<!-- DO NOT REMOVE -->

pjhampton commented:
This looks good so far 👍

mintsoft commented on share/goodie/conversions/conversions.js:
```diff
@@ -271,7 +274,8 @@ DDH.conversions = DDH.conversions || {};
                 { symbol: 'angstrom',   name: 'Angstrom' },
-                { symbol: 'mil',        name: 'Mil'}
+                { symbol: 'mil',        name: 'Mil'},
+                { symbol: 'nauticalmile', name: 'Nautical Mile' }
```
IIRC the symbol is technically NM

scottwarren replied:
While I work on the other lengths, what do we do when there is no agreed upon symbol (or where they are the same as another unit (eg `nm` is nanometer))?

scottwarren replied:
Nautical mile is a unit where there is [no agreed upon symbol](https://en.wikipedia.org/wiki/Nautical_mile#Unit_symbol)

pjhampton replied:
This is ok, it's only be displayed to the user and `nauticalmile` is only being used by math.js in this case. I think mintsoft is right though, we should add `NM` to the aliases in trigger.yml if it doesn't already exist. They are case sensitive AFAIK 😄

mintsoft replied:
That's fine then if it's internal only etc.

> where they are the same as another unit (eg nm is nanometer))?
units are always case sensitive so nm and NM are different units. For example `s` is seconds whilst `S` is Siemens (unit of conductance), we should attempt to keep this correct as much as possible. The behaviour we had previously was that units are case sensitive (s vs S) however names (SECONDS, SiEmEns) are case insensitive; that seemed to work reasonably well for all cases

scottwarren replied:
@mintsoft thank you, I was wondering about where some symbols are the same except for the case too. I'll keep working on this - I'll also try changing Nautical Mile to NM where it makes sense

pjhampton commented on share/goodie/conversions/triggers.yml:
```diff
@@ -319,6 +319,13 @@ type: length
 ---
 aliases:
+  - nmi
+  - NM
```
Ahh, you did. Yeah that's fine 👍

mintsoft commented on spec/conversions_spec.js:
```diff
@@ -113,5 +113,15 @@ describe("Conversion of Custom Units", function() {
+     */
+    
+    it("should convert nautical mile to meters", function() {
+       var converion = stripUnit(math.eval("1nauticalmile to meter"));
```
Typo: `conversion`

pjhampton commented:
I removed the merge conflicts @scottwarren. If you `g pull` you will be back on track 😄

pjhampton commented:
# pjhampton's sign off

* [ ] Micron
* [ ] Femtometer
* [ ] Attometer
* [ ] zeptometer
* [ ] Yoctometer
* [ ] Dekameter
* [ ] Hectometer
* [ ] Metameter
* [ ] Gigameter
* [ ] Terameter
* [ ] Petameter
* [ ] Exameter
* [ ] Zettameter
* [ ] Yottameter
* [ ] Nautical Mile
* [ ] Li
* [ ] Half Marathon
* [ ] Marathon
* [ ] Parsec
* [ ] Milliparsec
* [ ] Nanoparsec
* [ ] Picoparsec
* [ ] Kiloparsec
* [ ] Metaparsec
* [ ] Gigaparsec
* [ ] Teraparsec
* [ ] Astronomical Unit
* [ ] Light Year
* [ ] Leauge
* [ ] Fathom
* [ ] Smoot
* [ ] Cubit
* [ ] Furlong
* [ ] Megafurlong
* [ ] Beard Second

pjhampton commented on share/goodie/conversions/conversions.js:
```diff
@@ -74,6 +74,45 @@ DDH.conversions = DDH.conversions || {};
             {name: 'PB',    factor: '1000 TB'},
 
+            // CUSTOM LENGTH UNITS
+            {name: 'nauticalmile',     factor: '1852 metres'},
```
we're using the `er` spelling at the moment instead of `re`. These will need to be changed.
We can all chat about incorporating localization once all the unit PRs are in 👍

scottwarren replied:
Ahh yep - my bad. Australian here, forgot about the US spelling - sorry, will fix! :-)

pjhampton commented on share/goodie/conversions/conversions.js:
```diff
@@ -74,6 +74,45 @@ DDH.conversions = DDH.conversions || {};
+            {name: 'marathon',         factor: '42.195 kilometres'},
+            {name: 'li',               factor: '500 metres'},
+            {name: 'attometre',        factor: '0.000000000000001 metres'},
+            {name: 'femtometre',       factor: '10^-15 metres'},
```
Can't use exponent it seems. You will need to go to bing and actually get a unit like you have for `attometers`

scottwarren replied:
Ahh thank you, I was doing Wikipedia/Google searches to find each unit. I'll take a look at Bing :) :+1:

scottwarren replied:
Figured that I was using exponent wrong, and have fixed the creation of units. It was a big help having the front end running via [the Docker image](https://hub.docker.com/r/soleo/duckpan/)

pjhampton commented on share/goodie/conversions/conversions.js:
```diff
@@ -74,6 +74,45 @@ DDH.conversions = DDH.conversions || {};
+
+            // Need to find a defition for these
+            {name: 'milliparsec',      factor: ''},
+            {name: 'zeta',             factor: ''}
```
comma here... but you probably already know that 😉

scottwarren replied:
Thanks for spotting this! Just a mistake that would have thrown an error :D

pjhampton commented on share/goodie/conversions/conversions.js:
```diff
@@ -74,6 +74,45 @@ DDH.conversions = DDH.conversions || {};
+            {name: 'nauticalmile',     factor: '1852 metres'},
+            {name: 'halfmarathon',     factor: '21.0975 kilometres'},
+            {name: 'marathon',         factor: '42.195 kilometres'},
+            {name: 'li',               factor: '500 metres'},
```
It seems that `li` is already defined by math.js. You can kill this line. Keep it in the Units object though

scottwarren commented on share/goodie/conversions/conversions.js:
```diff
@@ -74,14 +74,52 @@ DDH.conversions = DDH.conversions || {};
+            {name: 'halfmarathon',     factor: '21.0975 kilometers'},
+            {name: 'marathon',         factor: '42.195 kilometers'},
+            {name: 'attometre',        factor: '0.000000000000001 meters'},
+            {name: 'femtometre',       factor: '10^-15 meters'},
```
Need to find actual number (can't use exponent) as per the [now collapsed comment](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/4181#discussion_r116559817)

scottwarren commented:
From what I understand I still have to:

* [ ] Add all units to the `triggers.yaml` including aliases for other common names
* [ ] Add "backend" tests to the `Conversions.t` file
* [ ] Fix tests that are breaking now in the `Conversions.t` file
* [ ] Fix frontend tests in the `conversions_spec.js`file
* [ ] Add (for ones I've missed) the front-end tests for unit conversions
* [ ] Find definition and implement for missing units `milliparsec` and `zeta` (does anyone know the answer to this?)

pjhampton commented:
Hey @scottwarren 
How are you getting on with this?

scottwarren commented:
Hey @pjhampton I haven't had a chance to look at this in the past week. I will take a look at adding the backend tests tonight, if that's ok? I understand things like this should be completed in a timely manner, and I apologise for this taking so long.

scottwarren commented:
@pjhampton I've fixed all of the front-end tests, and I've added a few backend tests. I've added all the units into the `triggers.yml`.

Math.js defines `li` as a [link](https://en.wikipedia.org/wiki/Link_(unit)) rather than a `chinese li`, so I've defined in in Math.js

scottwarren commented:
Sorry about the commits, there was a little while where I was pushing to see if the tests passed, I can now test locally. Happy to rebase/squash the commits :-)

scottwarren commented:
Usually I would resolve the conflicts myself, but I believe some of the conflicts are happening because of a refactor/cleanup of units, most of which I want to make sure remain as what's in `master` - can anyone have a look at this with me?

scottwarren commented:
I believe the conflicts are all sorted now :+1:

scottwarren commented:
For the review, I'd like to if possible have someone sign off these definitions, as there are a lot of different units, thus more likely to have a mistake:

* [ ] {name: 'nauticalmile',     factor: '1852 meters'},
* [ ] {name: 'halfmarathon',     factor: '21.0975 kilometers'},
* [ ] {name: 'marathon',         factor: '42.195 kilometers'},
* [ ] {name: 'attometre',        factor: '0.000000000000001 meters'},
* [ ] {name: 'femtometre',       factor: '0.0000000000000001 meters'},
* [ ] {name: 'zeptometre',       factor: '0.000000000000000000001 meters'},
* [ ] {name: 'yoctometre',       factor: '0.00000000000000000000001 meters'},
* [ ] {name: 'dekametre',        factor: '10 meters'},
* [ ] {name: 'hectometre',       factor: '100 meters'},
* [ ] {name: 'megametre',        factor: '1000 kilometers'},
* [ ] {name: 'gigametre',        factor: '1000000 kilometers'},
* [ ] {name: 'terametre',        factor: '1000000000 kilometers'},
* [ ] {name: 'petametre',        factor: '10000000000000000 meters'},
* [ ] {name: 'exametre',         factor: '10000000000000000000 meters'},
* [ ] {name: 'zettametre',       factor: '10000000000000000000000 meters'},
* [ ] {name: 'yottametre',       factor: '10000000000000000000000000 meters'},
* [ ] {name: 'parsec',           factor: '308567760000000000 meters'},
* [ ] {name: 'nanoparsec',       factor: '0.1 parsec'},
* [ ] {name: 'picoparsec',       factor: '0.000000000001 parsec'},
* [ ] {name: 'milliparsec',      factor: '0.001 parsec'},
* [ ] {name: 'kiloparsec',       factor: '1000 parsec'},
* [ ] {name: 'megaparsec',       factor: '1000 kiloparsec'},
* [ ] {name: 'gigaparsec',       factor: '1000 megaparsec'},
* [ ] {name: 'teraparsec',       factor: '1000 gigaparsec'},
* [ ] {name: 'astronomicalunit', factor: '149597870700 meters'},
* [ ] {name: 'lightyear',        factor: '9460730472580800 meters'},
* [ ] {name: 'league',           factor: '15840 feet'},
* [ ] {name: 'fathom',           factor: '6 feet'},
* [ ] {name: 'smoot',            factor: '1.7018 meters'},
* [ ] {name: 'cubit',            factor: '457.2 mm'},
* [ ] {name: 'furlong',          factor: '660 feet'},
* [ ] {name: 'megafurlong',      factor: '1000000 furlong'},
* [ ] {name: 'beardsecond',      factor: '10 nanometer'},
* [ ] {name: 'chineseli',        factor: '500 meters'},

scottwarren commented:
@pjhampton I haven't been able to find a `zeta` definition that I can define in `Math.js`, is this OK to leave out? I'm happy to raise another issue for that definition if you like?

Beside that, this is ready for PR.

With hindsight, I think we probably should have broken this into smaller tasks (to merge more often/earlier).

scottwarren commented:
Forgot to mention that there are a lot of commits due to how I was testing (coding locally, committing and pushing then running the tests on Codio). Happy to rebase these if it makes sense? On that note, do you guys use the `Squash` merge type when merging via Github?

scottwarren commented:
@mintsoft or @moollaza do you guys know who should be looking at this? Looks like @pjhampton is on vacation. I'm happy for this to wait until he gets back, but I got the impression that these tasks were important (and should be merged asap).

Sorry for bothering you guys. Thank you!

mintsoft commented:
@scottwarren I'm technically the maintainer for Conversions (although PJ has taken the lead on the JS rewrite) so I'll look after this

mintsoft commented:
@scottwarren 

>  Happy to rebase these if it makes sense? On that note, do you guys use the Squash merge type when merging via Github?

We squash on merge but if you want to squash to tidy up the history that's fine too.

I've checked through the unit definitions in your comment here https://github.com/duckduckgo/zeroclickinfo-goodies/pull/4181#issuecomment-304162928

The following are wrong:
* attometre - should be 1e-18
* femtometre - should be 1e-15
* yoctometre - should be 1e-24
* petametre - should be 1e15
* exametre - should be 1e18
* zettametre - should be 1e22
* yottameter - should be 1e25
* parsec - should be 30856776000000000 (~ 3.086e16)
* nanoparsec - should be e-9 parsec

The easiest way to check these is to log10 the number to check the order of magnitude

scottwarren commented:
Thanks @mintsoft I'll take a look at this tomorrow.

scottwarren commented:
@mintsoft, I've fixed the units you mentioned, except zettametre and yottameter, should they be:
```
Yottameter:
Math.log10(10000000000000000000000000); // => 25
// OR
Math.log10(1000000000000000000000000); // => 24

Zettameter:
Math.log10(10000000000000000000000); // => 22
// OR
Math.log10(1000000000000000000000); // => 21
```

Let me know if all good :) 👍

mintsoft commented:
@scottwarren Sorry for the delay, I missed this

Yotta should be 10^24
Zetta should be 10^21

mintsoft commented:
Also there are conflicts, can you rebase?

scottwarren commented:
@mintsoft all sorted 👍

mintsoft commented:
@pjhampton LGTM

pjhampton commented:
Unfortunately, the code base has evolved quite a bit since creating the original issue and the PR is quite out of sync with master. There is still valuable stuff here that we could consider an edge case. If you want, feel free to drop an up to date PR.
